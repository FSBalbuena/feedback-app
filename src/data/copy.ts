export const GatherInformation = {
  NEXT_BUTTON: "Next",
  PREV_BUTTON: "Previous",
  GET_FEEDBACK_BUTTON: "Get Feedback",
  TEXT_FIELD_LABEL: "Thoughts...",
};

export const Home = {
  START_BUTTON: "Start",
  TITLE:
    "Confidently give feedback that empowers others and strengthens relationships.",
  DESCRIPTION:
    "If you’re feeling unsure, anxious, or stuck about delivering constructive feedback, this is the place to find clarity, tools, and support. Together, we’ll turn your fears into confidence and your feedback into a positive force for growth.",
};

export const Feedback = {
  COPY_BUTTON: "Copy to clipboard",
  COPY_ALERT: "Copied to clipboard",
  COPY_ERROR: "Failed to copy to clipboard",
  RESTART_BUTTON: "Restart",
};

export const FeedbackSkeleton = {
  LOADING_MESSAGE: " Creating your feedback advice",
};
