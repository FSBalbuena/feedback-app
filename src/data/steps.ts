export const steps = [
  {
    id: "context",
    question:
      "What’s the situation or behavior you’d like to give feedback on?",
    hint: "Is this about a work project, a personal relationship, or something else?",
  },
  {
    id: "perspective",
    question:
      "What specifically did the person do or say that stood out to you?",
    hint: "Describe the actions or behaviors you noticed. Were they positive, negative, or mixed?",
  },
  {
    id: "impact",
    question: "How did this behavior affect you, the team, or the project?",
    hint: "Did it help or hinder progress? How did it make you feel?",
  },
  {
    id: "improvement",
    question: "What changes or improvements would you like to see?",
    hint: "Be specific—what could they do differently next time?",
  },
  {
    id: "strengths",
    question: "What did they do well that you’d like to acknowledge?",
    hint: "Even if there are areas for improvement, what positives can you highlight?",
  },
  {
    id: "tone",
    question: "How would you like to deliver this feedback?",
    hint: "Do you want it to be supportive, direct, or collaborative? Are there any sensitivities to keep in mind?",
  },
  {
    id: "outcome",
    question: "What do you hope to achieve with this feedback?",
    hint: "Is your goal to improve performance, strengthen communication, or something else?",
  },
];

export const stepsIDs = steps.map((step) => step.id);

export const questionsAndHints = steps.reduce((text, step) => {
  const { id, question, hint } = step;
  return text + `- ${id}: ${question} --> ${hint}\n`;
}, "");
