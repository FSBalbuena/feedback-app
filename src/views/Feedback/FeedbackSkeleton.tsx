import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import Skeleton from "@mui/material/Skeleton";
import { FeedbackSkeleton as copyText } from "@/data/copy";

const FeedbackSkeleton = () => (
  <Box width={`90vw`} maxWidth={800}>
    <Typography variant="body1" gutterBottom>
      {copyText.LOADING_MESSAGE}
    </Typography>
    <Typography variant="h2" gutterBottom>
      <Skeleton width={"60%"} />
    </Typography>
    <Typography variant="body1" gutterBottom>
      <Skeleton />
    </Typography>
    <Typography variant="body1" gutterBottom>
      <Skeleton />
    </Typography>
    <Typography variant="body1" gutterBottom>
      <Skeleton />
    </Typography>
    <Typography variant="body1" gutterBottom>
      <Skeleton />
    </Typography>
    <Typography variant="body1" gutterBottom>
      <Skeleton />
    </Typography>
    <Typography variant="h3" gutterBottom>
      <Skeleton width={"60%"} />
    </Typography>
    <Typography variant="body1" gutterBottom>
      <Skeleton />
    </Typography>
    <Typography variant="body1" gutterBottom>
      <Skeleton />
    </Typography>
    <Typography variant="body1" gutterBottom>
      <Skeleton />
    </Typography>
  </Box>
);

export default FeedbackSkeleton;
