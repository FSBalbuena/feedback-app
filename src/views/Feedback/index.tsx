import { useState } from "react";
import Markdown from "react-markdown";
import Grid from "@mui/material/Grid2";
import Tooltip from "@mui/material/Tooltip";
import IconButton from "@mui/material/IconButton";
import Snackbar from "@mui/material/Snackbar";
import Alert from "@mui/material/Alert";
import ContentCopyIcon from "@mui/icons-material/ContentCopy";
import RestartAltIcon from "@mui/icons-material/RestartAlt";
import { Feedback as copyText } from "@/data/copy";

type Props = {
  feedbackAdvice: string;
  onRestart: () => void;
};
function Feedback({ feedbackAdvice, onRestart }: Props) {
  const [open, setOpen] = useState(false);

  const handleCopy = async () => {
    try {
      await navigator.clipboard.writeText(feedbackAdvice);
      setOpen(true); // Show confirmation snackbar
    } catch (err) {
      console.error(copyText.COPY_ERROR, err);
    }
  };

  const handleClose = () => {
    setOpen(false);
  };
  return (
    <Grid container maxWidth={800} padding={4}>
      <Grid size={12}>
        <Markdown>{feedbackAdvice}</Markdown>
      </Grid>

      <Grid size={12} container justifyContent={"space-around"}>
        <Tooltip title={copyText.COPY_BUTTON}>
          <IconButton onClick={handleCopy}>
            <ContentCopyIcon />
          </IconButton>
        </Tooltip>
        <Tooltip title={copyText.RESTART_BUTTON}>
          <IconButton onClick={onRestart}>
            <RestartAltIcon />
          </IconButton>
        </Tooltip>
      </Grid>
      <Snackbar
        open={open}
        autoHideDuration={2000}
        onClose={handleClose}
        anchorOrigin={{ vertical: "top", horizontal: "right" }}
      >
        <Alert onClose={handleClose} severity="success">
          {copyText.COPY_ALERT}
        </Alert>
      </Snackbar>
    </Grid>
  );
}

export default Feedback;
