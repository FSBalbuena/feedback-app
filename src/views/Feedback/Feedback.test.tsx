import Feedback from ".";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { Feedback as copyText } from "@/data/copy";

const feedbackAdvice = `# Feedback Advice \n
testing markdown with *snapshot testing*\n
1. test ordered list\n
1.1 test ordered list 2\n
and some tips:\n
* tip 1
* tip 2`;

const onRestart = jest.fn();

const DEFAULT_PROPS = {
  feedbackAdvice,
  onRestart,
};

describe("Feedback", () => {
  describe("on Render", () => {
    it("should render", () => {
      const { container } = render(<Feedback {...DEFAULT_PROPS} />);
      expect(container).toMatchSnapshot();
    });
  });
  describe("interactive", () => {
    describe("when user clicks on restart", () => {
      it("should call onRestart", () => {
        render(<Feedback {...DEFAULT_PROPS} />);
        const button = screen.getByRole("button", {
          name: copyText.RESTART_BUTTON,
        });
        expect(button).toBeInTheDocument();
        expect(onRestart).not.toHaveBeenCalled();

        fireEvent.click(button as Element);
        expect(onRestart).toHaveBeenCalled();
      });
    });

    describe("when user clicks on copy", () => {
      it("should fired an alert", () => {
        render(<Feedback {...DEFAULT_PROPS} />);
        const button = screen.getByRole("button", {
          name: copyText.COPY_BUTTON,
        });
        expect(button).toBeInTheDocument();
        const alert = screen.queryByRole("alert");
        expect(alert).toBeNull();

        fireEvent.click(button as Element);
        waitFor(() => {
          const alert = screen.queryByRole("alert");
          expect(alert).toBeInTheDocument();
          expect(alert).toHaveTextContent(copyText.COPY_ALERT);
        });
      });
    });
  });
  describe("behavior", () => {
    describe("when there is an error on copy", () => {
      beforeAll(() => {
        Object.assign(navigator, {
          clipboard: {
            writeText: jest.fn().mockImplementation(() => {
              throw new Error("Clipboard write failed");
            }),
          },
        });
      });

      afterAll(() => {
        jest.restoreAllMocks();
      });
      it("should not fired an alert", () => {
        render(<Feedback {...DEFAULT_PROPS} />);
        const button = screen.getByRole("button", {
          name: copyText.COPY_BUTTON,
        });
        expect(button).toBeInTheDocument();
        const alert = screen.queryByRole("alert");
        expect(alert).toBeNull();

        fireEvent.click(button as Element);
        waitFor(() => {
          const alert = screen.queryByRole("alert");
          expect(alert).toBeNull();
        });
      });
    });
  });
});
