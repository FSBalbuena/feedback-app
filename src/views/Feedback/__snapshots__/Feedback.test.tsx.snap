// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback on Render should render 1`] = `
<div>
  <div
    class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-18l5w1l-MuiGrid2-root"
  >
    <div
      class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-1j0u08r-MuiGrid2-root"
    >
      <div>
        # Feedback Advice 

testing markdown with *snapshot testing*

1. test ordered list

1.1 test ordered list 2

and some tips:

* tip 1
* tip 2
      </div>
    </div>
    <div
      class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-n3xlqn-MuiGrid2-root"
    >
      <button
        aria-label="Copy to clipboard"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ContentCopyIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2m0 16H8V7h11z"
          />
        </svg>
      </button>
      <button
        aria-label="Restart"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="RestartAltIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 5V2L8 6l4 4V7c3.31 0 6 2.69 6 6 0 2.97-2.17 5.43-5 5.91v2.02c3.95-.49 7-3.85 7-7.93 0-4.42-3.58-8-8-8m-6 8c0-1.65.67-3.15 1.76-4.24L6.34 7.34C4.9 8.79 4 10.79 4 13c0 4.08 3.05 7.44 7 7.93v-2.02c-2.83-.48-5-2.94-5-5.91"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
